src/v/iceberg/tests/merge_append_action_test.cc

```diff
@@ -260,6 +280,92 @@ TEST_F(MergeAppendActionTest, TestMergeByBytes) {
+    const size_t files_per_man = 2;
+    const size_t rows_per_man = 25;
+    transaction tx(create_table());
+    // Repeatedly add new data files up to the merge count threshold. First half
+    // of the files will be with the old schema.
+    for (size_t i = 0; i < num_to_merge_at / 2; i++) {
+        const auto expected_snapshots = i + 1;
+        const auto expected_manifests = expected_snapshots;
+        merge_append_and_check(
+          tx,
+          create_data_files(tx.table(), "foo", files_per_man, rows_per_man),
+          expected_snapshots,
+          expected_manifests);
+    }
+
+    // Promote int partition column to long.
+    {
+        auto orig_type = tx.table().schemas.at(0).schema_struct.copy();
+        auto new_type = orig_type.copy();
+        bool found = false;
+        for (auto& field : new_type.fields) {
+            if (field->name == "bar") {
+                field->type = long_type{};
+                found = true;
+            }
+        }
+        ASSERT_TRUE(found);
+
+        auto compat_res = evolve_schema(orig_type, new_type);
+        ASSERT_FALSE(compat_res.has_error());
+
+        auto res = tx.set_schema(iceberg::schema{
+                                   .schema_struct = std::move(new_type),
+                                   .schema_id = iceberg::schema::unassigned_id,
+                                   .identifier_field_ids = {},
+                                 })
+                     .get();
+        ASSERT_FALSE(res.has_error()) << res.error();
+    }
+
+    // Add the remaining files, now partition column value is long.
+    for (size_t i = num_to_merge_at / 2; i < num_to_merge_at; i++) {
+        const auto expected_snapshots = i + 1;
+        const auto expected_manifests = expected_snapshots;
+        merge_append_and_check(
+          tx,
+          create_data_files(
+            tx.table(), "foo", files_per_man, rows_per_man, long_value{42}),
+          expected_snapshots,
+          expected_manifests);
+    }
+
+    // At the merge threshold, we expect the latest snapshot contains a merged
+    // manifest. All manifests should be successfully merged, even though the
+    // type of the partition column is not the same.
+    auto res = tx.merge_append(
+                   io,
+                   create_data_files(
+                     tx.table(), "foo", files_per_man, rows_per_man))
+                 .get();
+    ASSERT_FALSE(res.has_error()) << res.error();
+    const auto& table = tx.table();
+    ASSERT_TRUE(table.snapshots.has_value());
+    ASSERT_EQ(table.snapshots.value().size(), num_to_merge_at + 1);
+
+    // Validate that the latest snapshot indeed contains a single manifest.
+    auto latest_mlist_path = table.snapshots.value().back().manifest_list_path;
+    auto latest_mlist = io.download_manifest_list(latest_mlist_path).get();
+    ASSERT_TRUE(latest_mlist.has_value());
+    ASSERT_EQ(latest_mlist.value().files.size(), 1);
```

Comment:
q (curious): what would have happened here before these changes?

Reply:
Previous version threw an exception (IIRC when comparing values of different types for calculating partition summaries)

Reply:
Ah that's a bummer -- I've been trying to avoid mixing exceptions and result types and guess I missed that. Can that still throw as is?

Reply:
Well, comparisons are under a try-catch now as well (`merge_append_action.cc:501`), so exceptions there should be converted to error codes as well